SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/VisibilityTest/VisibilityTest.cpp

```diff
@@ -0,0 +1,344 @@
+					{
+						NetTest->FinishStep();
+					}
+				}
+			}
+		});
+	}
+
+	{   // Step 4 - Make sure that the Client moved to the correct location.
+		AddStep(TEXT("ClientCheckFirstMovement"), FWorkerDefinition::Client(1), nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime)
+		{
+			ASpatialFunctionalTestFlowController* FlowController = GetLocalFlowController();
+			checkf(FlowController, TEXT("Can't be running test without valid FlowControl."));
+
+			APlayerController* PlayerController = Cast<APlayerController>(FlowController->GetOwner());
+			ATestMovementCharacter* PlayerCharacter = Cast<ATestMovementCharacter>(PlayerController->GetPawn());
+			if (IsValid(PlayerCharacter))
+			{
+				if (PlayerCharacter->GetActorLocation().Equals(CharacterRemoteLocation, 50.0f))
+				{
+					FinishStep();
+				}
+			}
+
+		}, 5.0f);
+	}
+
+	{	// Step 5 - Observe if the test actor has been replicated to the client 1.
+		AddStep(TEXT("ClientCheckReplicatedActorsBeforeActorHidden"), FWorkerDefinition::Client(1), nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime){
+			int Counter = 0;
+			int ExpectedReplicatedActors = 0;
+			for (TActorIterator<AReplicatedVisibilityTestActor> Iter(NetTest->GetWorld()); Iter; ++Iter)
+			{
+				Counter++;
+			}
+			if (Counter == ExpectedReplicatedActors)
+			{
+				NetTest->AssertEqual_Int(Counter, ExpectedReplicatedActors, TEXT("Number of TestHiddenActors in client 1 world"), NetTest);
+				NetTest->FinishStep();
+			}
+		}, 5.0f);
+	}
+
+	{	// Step 6 - Observe if the test actor has been replicated to the client 2.
+		AddStep(TEXT("ClientCheckReplicatedActorsBeforeActorHidden"), FWorkerDefinition::Client(2), nullptr, nullptr, [this](ASpatialFunctionalTest* NetTest, float DeltaTime) {
+			int Counter = 0;
+			int ExpectedReplicatedActors = 1;
+			for (TActorIterator<AReplicatedVisibilityTestActor> Iter(NetTest->GetWorld()); Iter; ++Iter)
+			{
+				Counter++;
+			}
+			if (Counter == ExpectedReplicatedActors)
+			{
+				NetTest->AssertEqual_Int(Counter, ExpectedReplicatedActors, TEXT("Number of TestHiddenActors in client 2 world"), NetTest);
+				NetTest->FinishStep();
+			}
+		}, 5.0f);
+	}
+
+	{	// Step 7 - Server Set Actor Hidden.
+		AddStep(TEXT("ServerSetActorHidden"), FWorkerDefinition::Server(1), nullptr, [this](ASpatialFunctionalTest* NetTest) {
+			int Counter = 0;
+			int ExpectedReplicatedActors = 1;
+			for (TActorIterator<AReplicatedVisibilityTestActor> Iter(NetTest->GetWorld()); Iter; ++Iter)
+			{
+				Counter++;
+				Iter->SetHidden(true);
+				NetTest->AssertEqual_Bool(Iter->IsHidden(), true, TEXT("TestHiddenActors is Hidden in the server world"), NetTest);
+			}
+			NetTest->AssertEqual_Int(Counter, ExpectedReplicatedActors, TEXT("Number of TestHiddenActors in server world"), NetTest);
+			NetTest->FinishStep();
+		});
+	}
+
+	{	// Step 8 - Observe if the test actor has been replicated to the server.
```

Comment:
Is this step necessary? Wouldn't the server always have the actor?

Reply:
Just thought it will be good to double-check if Server has the actor as I changed the QBIs, so you confirm that only clients got affected by this change

Reply:
The server should always have the actor locally since the actor has not been deleted and spatial won't tell it to delete the actor in the case it has been marked hidden.